### DISCUSSION TITLE
关于Border.Background采用ImageBrush时,频繁刷新绑定值, 资源不释放的问题

### DISCUSSION THREAD

**Discussion by hongyanfengzi** [BYJRK/DotNet-Discussions#65] [WPF 开发]
<Border Grid.Row="0" Width="auto" Margin="0,10">
                            <Border.Background>
                                <ImageBrush
                                    ImageSource="{Binding DetBackground}" />
                            </Border.Background>
</Border>

DetBackground 是一个其他程序集的资源图片, 是一个string类型的地址引用
 string DetBackground = "/dllName;Component/images/pic_det_bk_empty.png" //背景图标
当我以比较高的频率修改DetBackground 值时, 内存会一直增长, 得不到释放

**Comment by hongyanfengzi** [comment]
` <Border Grid.Row="0" Width="auto" Margin="0,10">
                            <Border.Background>
                                <ImageBrush
                                    ImageSource="{Binding DetBackground}" />
                            </Border.Background>
</Border>
`

**Comment by BYJRK** [comment]
能否大致说一下你是怎样“以较高频率修改DetBackground值”的？

**Reply by BYJRK** [reply]
这里，你绑定的是字符串，有可能是因为string经由TypeConverter转为ImageSource后，每次都要声明新的ImageSource对象？

**Reply by hongyanfengzi** [reply]
> 能否大致说一下你是怎样“以较高频率修改DetBackground值”的？

每过几秒钟会读取一个硬件设备的状态, 状态变了就会修改这个背景值

**Reply by hongyanfengzi** [reply]
> 这里，你绑定的是字符串，有可能是因为string经由TypeConverter转为ImageSource后，每次都要声明新的ImageSource对象？

这里是因为绑定了string值, 每次都要声明新的ImageSource, 我也发现了这个问题, 但是修改的方案我没有想到合适的

**Reply by BYJRK** [reply]
不要绑定字符串，而是绑定一个enum或者int等表示当前状态的值，然后写一个Style，里面用DataTrigger来决定使用哪张图片

**Reply by hongyanfengzi** [reply]
> 不要绑定字符串，而是绑定一个enum或者int等表示当前状态的值，然后写一个Style，里面用DataTrigger来决定使用哪张图片

我本身是有一个枚举状. DataTrigger 这个我还真不了解, 我去试一下, 感谢啊

**Reply by hongyanfengzi** [reply]
> 不要绑定字符串，而是绑定一个enum或者int等表示当前状态的值，然后写一个Style，里面用DataTrigger来决定使用哪张图片

我本来是想写一个converter 转换一下绑定枚举值, 返回静态的imageScore, 但是在converter里面我一直拿不到别的程序集的资源图片

**Reply by BYJRK** [reply]
你在xaml里面写的，直接写在 Uri 里面就行了，比如

`new Uri("/dllName;Component/images/pic_det_bk_empty.png", UriKind.RelativeOrAbsolute)`

**Reply by hongyanfengzi** [reply]
> 你在xaml里面写的，直接写在 Uri 里面就行了，比如
> 
> `new Uri("/dllName;Component/images/pic_det_bk_empty.png", UriKind.RelativeOrAbsolute)`

好的, 我试一下, 那converter的方式, 和 DataTrigger , 哪种更合适一些的呢

**Reply by BYJRK** [reply]
用 Trigger 更好，Converter 除非你共享资源，比如写一个静态对象
